Source/NSDictionary.m

```diff
@@ -958,8 +959,13 @@ - (NSArray *)keysSortedByValueUsingComparator: (NSComparator)cmptr
 - (NSArray *)keysSortedByValueWithOptions: (NSSortOptions)opts
 			  usingComparator: (NSComparator)cmptr
 {
-  return [[self allKeys] sortedArrayWithOptions: opts
-				usingComparator: cmptr];
+  NSArray* sortedValues = [[self allValues] sortedArrayWithOptions: opts usingComparator: cmptr];
+  NSArray* noDuplicates = [[NSOrderedSet orderedSetWithArray:sortedValues] array];
+  NSMutableArray* result = [[NSMutableArray alloc] initWithCapacity:[sortedValues count]];
+  for (NSObject* value in noDuplicates) {
```

Comment:
All current compiler are able to process this modern for loop, still we prefer to use the macros FOR_IN and END_FOR_IN.
Although somebody should check the support of gcc in GSFastEnumeration.h as you did not get a compiler warning for this.

Reply:
Fred, I added fixes for the issues you spotted, plus I tidied to match coding style and fixed a memory leak.